src/sentry/api/endpoints/project_stacktrace_link.py

```diff
@@ -271,19 +340,43 @@ def get(self, request: Request, project: Project) -> Response:
                 )
                 scope.set_tag("codecov.enabled", codecov_enabled)
                 if codecov_enabled:
+                    # Get Codecov data
                     try:
-                        lineCoverage, codecovUrl = get_codecov_data(
-                            repo=current_config["config"]["repoName"],
-                            service=current_config["config"]["provider"]["key"],
-                            branch=current_config["config"]["defaultBranch"],
-                            path=current_config["outcome"]["sourcePath"],
+                        # Get ref
+                        has_error_commit = bool(ctx.get("commit_id"))
+                        ref = ctx["commit_id"] if has_error_commit else None
+                        fetch_commit_sha = features.has(
+                            "organizations:codecov-commit-sha-from-git-blame",
+                            project.organization,
+                            actor=request.user,
                         )
-                        if lineCoverage and codecovUrl:
-                            result["codecov"] = {
-                                "lineCoverage": lineCoverage,
-                                "coverageUrl": codecovUrl,
-                                "status": 200,
-                            }
+                        line_no = ctx.get("line_no")
+                        if not has_error_commit and fetch_commit_sha:
+                            self.get_ref(
+                                integrations,
+                                project.organization_id,
+                                int(line_no) if line_no else 0,
+                                filepath,
+                                current_config["repository"],
+                                current_config["config"]["defaultBranch"],
+                            )
+
+                        # Call codecov API if codecov-commit-sha-from-git-blame flag is enabled
+                        # or getting ref from git blame was successful
+                        if not fetch_commit_sha or ref:
+                            lineCoverage, codecovUrl = get_codecov_data(
+                                repo=current_config["config"]["repoName"],
+                                service=current_config["config"]["provider"]["key"],
+                                ref=ref if ref else current_config["config"]["defaultBranch"],
+                                ref_type="sha" if ref else "branch",
+                                path=current_config["outcome"]["sourcePath"],
+                            )
+                            if lineCoverage and codecovUrl:
+                                result["codecov"] = {
+                                    "lineCoverage": lineCoverage,
+                                    "coverageUrl": codecovUrl,
+                                    "status": 200,
+                                }
```

Comment:
I think we could move this entire block into something like `fetch_codecov_data()` and have that method return a dictionary for the result fields, which you can append to result via `result.update(codecov_results)`. If that starts getting messy, dw and we can clean this up more later!

Reply:
Thanks! I moved the whole block into a function